Dark launch: Fix 1+N queries in third party auth config fetching

timmc-edx opened the pull request:
Dark launch comparison of a fix for the 1+N issue on the login page. Old implementation is still run and the results used, but if there are discrepancies with the new implementation, logs metrics and debugging info to New Relic.

This should also fix an issue with configs shadowing each other across sites, but we only have configs for one site in Production, so I'm not expecting the comparison to trip over that.

feanil commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -30,19 +31,17 @@ def _enabled_providers(cls):
-            provider = OAuth2ProviderConfig.current(oauth2_backend_name)
+        oauth2_backends = OAuth2ProviderConfig.objects.prefetch_related('site').all()
+        for provider in unique_everseen(oauth2_backends, key=lambda p: p.backend_name):
             if provider.enabled_for_current_site and provider.backend_name in _PSA_OAUTH2_BACKENDS:
```
Since we're only ever doing this for the site we're on, can we do the reverse and get all the providers where the site is the current site?

If I'm reading this right, it looks like a provider is mapped to at most one site:
https://github.com/edx/edx-platform/blob/0f5c474f8f49e7dbc1a7d60e15f5409a29dfe69f/common/djangoapps/third_party_auth/models.py#L138

timmc-edx replied:
Agreed that that would make more sense. For that matter, why not add `site`, `enabled`, and `backend_name` to the query and skip the entire guard clause that's inside the for loop?

I was being cautious because I'm not sure why it wasn't written that way in the first place... but maybe there's not a good reason. :-)

feanil replied:
Yea, if we find one later, we can always undo the change and put in the good reason as a comment.  Let's just make this a more sensible query.

timmc-edx replied:
Curiously, most of the tests pass if I do the below, including one of the two that was failing before—but the other one (`test_provider.py:38`) is still failing as if the query is being cached.

```py
        site = Site.objects.get_current(get_current_request())
        oauth2_backends = OAuth2ProviderConfig.objects.filter(enabled=True, site=site, backend_name__in=_PSA_OAUTH2_BACKENDS)
        for provider in unique_everseen(oauth2_backends, key=lambda p: p.backend_name):
            yield provider
```

Failure:

```py
    def test_runtime_configuration(self):
        self.configure_google_provider(enabled=True)
        enabled_providers = provider.Registry.enabled()
        self.assertEqual(len(enabled_providers), 1)
        self.assertEqual(enabled_providers[0].name, "Google")
        self.assertEqual(enabled_providers[0].get_setting("SECRET"), "opensesame")
    
        self.configure_google_provider(enabled=False)
        enabled_providers = provider.Registry.enabled()
>       self.assertEqual(len(enabled_providers), 0)
E       AssertionError: 1 != 0

common/djangoapps/third_party_auth/tests/test_provider.py:38: AssertionError
```

timmc-edx replied:
OK, I think I see what's happening here. The test configures a provider that is enabled, then configures one that is disabled, both with the same backend name, so that we end up with two objects in the DB. The code on master is supposed to pick up _only the most recent object_, and then discard it if it is disabled. That's why it's not part of the query.

(But then I don't understand why site wasn't part of the query! Surely we'd want to partition that append-to-override behavior by site.)

timmc-edx replied:
(And the reason my code was failing is because it was getting the _first_ row with a given slug or backend or key, not the _last_.)

feanil replied:
So this is because it's backed by config models right?  The config model table is essentially keeping a history of all the ways a setting has been set. So to get the correct setting, we want to get the latest setting for any given name and treat it as the source of truth.  So if the latest version is disabled, then we treat the value as disabled.  It seems like the test is valid.

timmc-edx replied:
Yeah, and my iterator was skipping the disabled one. I think the existing code has a bug around handling of sites, though; if two providers are registered for the same `backend_name` but different sites, the existing code makes the more recent one shadow the other.

nasthagiri commented:
Curious: Is there a ticket for this work? I'm wondering if we are investing time in preemptive optimization here.

schenedx commented:
@timmc-edx What is the context here? What problem did you see that prompt this change? The change is not really obvious to me it improves things. I read your PR description and comments and couldn't make good sense out of them.

timmc-edx commented:
Sorry, yes, this is https://openedx.atlassian.net/browse/ARCHBOM-1139 -- we're seeing an inordinate number of SELECTs on `django_site` driven just by loading the login page.

schenedx commented:
@timmc-edx Great. With that ticket, now I know what you are trying to do. I'll need to review again. Thanks

timmc-edx commented:
At this point, all I can say for sure is that there's something I'm missing around how `current`, `KEY_FIELDS`, and `site_id` are interacting with the various subclasseses of ProviderConfig in provider.py. :-)

(The Slack thread you were tagged in contains some of my stream of consciousness, so maybe that's already obvious by now.)

schenedx commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -29,21 +30,43 @@ def _enabled_providers(cls):
+            # Get the most recent record in each partition
+            return unique_everseen(configs.order_by('-change_date'), partition_key)
+
+        site = Site.objects.get_current(get_current_request())
```
It's easier to read if we have this statement before the `current_partitioned` function.

timmc-edx replied:
Agreed, although my intention would be to move `current_partitioned` out to django-config-models—or more likely, to change `ConfigurationModel.current` to accept a QuerySet or a list of additional filters. It's actually not totally clear if this is necessary at all, though; I'm looking into whether we can just add `'site'` to the `KEY_FIELDS` of each model, which should fix the bug. (Might not help with the 1+N, though!)

timmc-edx replied:
(Ah... I think I do want to add `'site'` to `KEY_FIELDS`, but it wouldn't be sufficient to address 1+N.)

schenedx commented on common/djangoapps/third_party_auth/tests/test_provider.py:
```diff
@@ -147,6 +147,27 @@ def test_provider_disabled_for_mismatching_site(self):
         self.assertEqual(prov.enabled_for_current_site, False)
 
+    @with_site_configuration(SITE_DOMAIN_A)
+    def test_providers_with_same_key_independent_across_sites(self):
```
I believe we can have another test case where each `site` will have both oauth_provider set and SAML_provider set. Then the return should be the latest oauth_provider and latest saml provider in a list.

timmc-edx replied:
Good idea.

schenedx commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -29,21 +30,43 @@ def _enabled_providers(cls):
+                if config.enabled:
+                    yield config
+
+        for provider in enabled_in_site(OAuth2ProviderConfig, {'backend_name__in': _PSA_OAUTH2_BACKENDS}):
```
Can you please add more comments for this loop? Like `First, we should get all the ID providers using oauth2`

schenedx commented on common/djangoapps/third_party_auth/provider.py:
```diff
@@ -29,21 +30,43 @@ def _enabled_providers(cls):
+
+        for provider in enabled_in_site(OAuth2ProviderConfig, {'backend_name__in': _PSA_OAUTH2_BACKENDS}):
+            yield provider
+        if SAMLConfiguration.is_enabled(site, 'default'):
```
Please add another comment before this if statement like `Then we get all the ID providers using SAML`

timmc-edx commented:
Addressed some feedback. I'm also going to see about:

- Reconciling this with `get_enabled_by_backend_name` lower in the same file
- Adding `'site_id'` to `KEY_FIELDS` -- this would be a larger change, but would make all of this structurally safer, and the changes to django-config-models would be simpler (no need for a filter argument in `current_all`). Might turn into a separate PR?

timmc-edx commented:
I've changed it to be a dark launch for now, so that the old implementation is still used, but the new code is run under a try/catch and any failures or result mismatches are sent to New Relic.

My intention is to open a separate PR to put `current_all` into django-config-models, then update this PR, but I'm keeping it here until just before merge for the sake of a faster review cycle. Note that it still lacks caching; if I were able to add site_id to the KEY_FIELDS of the models, I could get rid of the `filters` param and add caching, but lacking that I don't think it makes sense.
